src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -32,20 +33,33 @@ public static DynamicAnalysisInjector TryCreate(MethodSymbol method, BoundStatem
         {
             // Do not instrument implicitly-declared methods, except for constructors.
             // Instrument implicit constructors in order to instrument member initializers.
-            if (!method.IsImplicitlyDeclared || method.IsImplicitConstructor)
+            if (method.IsImplicitlyDeclared && !method.IsImplicitConstructor)
             {
-                MethodSymbol createPayload = GetCreatePayload(methodBodyFactory.Compilation, methodBody.Syntax, diagnostics);
+                return null;
+            }
 
-                // Do not instrument any methods if CreatePayload is not present.
-                // Do not instrument CreatePayload if it is part of the current compilation (which occurs only during testing).
-                // CreatePayload will fail at run time with an infinite recursion if it Is instrumented.
-                if ((object)createPayload != null && !method.Equals(createPayload))
-                {
-                    return new DynamicAnalysisInjector(method, methodBody, methodBodyFactory, createPayload, diagnostics, debugDocumentProvider, previous);
-                }
+            // Do not instrument methods marked with or in scope of ExcludeFromCodeCoverageAttribute.
+            if (IsExcludedFromCodeCoverage(method))
+            {
+                return null;
+            }
+
+            MethodSymbol createPayload = GetCreatePayload(methodBodyFactory.Compilation, methodBody.Syntax, diagnostics);
+
+            // Do not instrument any methods if CreatePayload is not present.
+            if (createPayload is null)
+            {
+                return null;
             }
 
-            return null;
+            // Do not instrument CreatePayload if it is part of the current compilation (which occurs only during testing).
```

Comment:
That's the only situation in which it would be being compiled. CreatePayload is part of a compilation with instrumentation only when we are building the dynamic analysis runtime with instrumentation, which occurs only during our own testing.